README.md
# Endoor

Endoor is a powerful penetration testing tool which supports network
reconnaissance and network infiltration.  Endoor is intended to be run on a
small device with 2 Ethernet interfaces, such as a Raspberry Pi, and shall be
plugged in between the victim's network and a node of the network (e.g. a
computer or printer).  Endoor transparently switches all frames back and forth
between its 2 network interfaces.  After a learning phase you can access the
network. Endoor transparently reuses the MAC address and the IP address of the
victim's node thereby being almost undetectable and bypassing L2 protection
mechanisms such as MAC filtering or 802.1x authentication.

The following diagram depicts its general use-case.

```
                        outside           inside
                       interface         interface
                           |                 |
( victim's network ) <----> ( Endoor device ) <----> ( victim's node )
                                    ^
                                    |
                           ( mobile network )
                                    |
                                    |
                               ( attacker )
```

# How to use Endoor

As already said it is intended to be installed on a small device (the endoor
device) on which endoor is run. The endoor device must have 2 Ethernet
interfaces and any 3rd interface through which you (the pentester) can access
the device. In real scenarios this most probably will be through a mobile
network but in lab environments you could access it through Wifi as well.

Once setup properly (see **Preparation** below) just connect it in between a
computer (or other device) and the uplink. Then ssh to it through your access
network and attack ;)

# Preparation

The 2 network interfaces are called "inside" and "outside" interface where
inside is the interface to the node and outside is the interface to the network
(see diagram above).
These 2 network interfaces must be unconfigured and silent. That means set it
up in such a way that there are no IP addresses configured and the DHCP client
has to be disabled.

If you use a Raspberry Pi, edit the file `/etc/dhcpcd.conf` and add the
following 2 lines (assuming eth0 and eth1 are the inside and the outside
interface):
```
denyinterfaces eth0
denyinterfaces eth1
```

I also suggest to disable any service accept sshd. Disable IPV6
auto-configuration and acception of router advertisements. This is necessary to
not reveal your device (see `files/osprep` of the source package).
```
sysctl -w net.ipv4.conf.all.forwarding=0
sysctl -w net.ipv6.conf.all.forwarding=0
sysctl -w net.ipv6.conf.eth0.autoconf=0
sysctl -w net.ipv6.conf.eth1.accept_ra=0
```

In a next step copy the endoor package to it, unpack, compile, and install it.
```
tar xfz endoor-1.0.tar.gz
cd endoor-1.0
./configure
make
sudo make install
```

You can now start endoor directly on the command line:
```
endoor -i eth0 -o eth1
```

It will immediately start to forward frames and will open a command line
interface. The command line is mainly for debugging. Use `help` to get a list
of commands.

Endoor will learn addresses from the network and tries to find the clients IP
address and MAC address as well as the MAC address of the router in the
network. Once it found it it will open and configure a tunnel device (tun0).
The victim's device's IP address will be set locally on the tunnel device.

From the address table in the command line you may have learned from the IP
range of the victim's network. Let's assume all internal addresses are in the
range of 192.168.0.0/16.

So open a new shell and add a route to the tunnel device:
```
sudo ip route add 192.168.0.0/16 dev tun0
```

From this point you can access any address within this range from the endoor
device, e.g.:
```
ping 192.168.17.23
```

Endoor will send all packets coming in locally through the tunnel device the
victim's network with a source address of the victim's node's IP an MAC
address.
Internally, endoor builds up a state table. Incoming packets from the victim's
network are matched against the state table and all packets with a proper state
are then diverted back to the tunnel device.

Since the states depend on the higher layer protocols, only protocols which are
implemented within Endoor are supported.
Currently this is TCP, UDP, and ICMP echo requests.

But that's more than enough for now, you can e.g. `nmap` through it ;)


# Detectability

## Physical detection

## Bridge transparency

## Self-revealed

## Timing

## State collisions

## Switch port state

## Traffic analysis



ChangeLog
* 2022/09/12 1st working version.



src/cli.c
/* Copyright 2022 Bernhard R. Fischer.
 *
 * This file is part of Endoor.
 *
 * Endoor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 *
 * Endoor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Endoor. If not, see <http://www.gnu.org/licenses/>.
 */

/*! \file cli.c
 * This file contains the code for the cli.
 *
 *  \author Bernhard R. Fischer <bf@abenteuerland.at>
 *  \date 2022/09/19
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <pthread.h>
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_SOCKET_H
#include <sys/socket.h>
#endif
#ifdef HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif
#ifdef HAVE_LINUX_IF_PACKET_H
#include <linux/if_packet.h>
#endif
#ifdef HAVE_NET_IF_H
#include <net/if.h>
#endif
#ifdef HAVE_SYSLOG_H
#include <syslog.h>
#endif
#ifdef HAVE_ARPA_INET_H
#include <arpa/inet.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif

#include "endoor.h"
#include "pcap.h"
#include "log.h"
#include "bridge.h"
#include "tun.h"
#include "estring.h"

#define SNAPLEN 4096
#define MACTABLESIZE 1024
#define STATETABLESIZE 16384


int data_to_hex(char *dst, const char *buf, int buflen)
{
   const char *hex = "0123456789abcdef";
   int len;

   // safety check
   if (buflen <= 0 || buf == NULL || dst == NULL)
      return 0;

   for (len = 0; buflen > 0; buflen--, buf++, len += 3)
   {
      *dst++ = hex[(*buf) >> 4 & 0xf];
      *dst++ = hex[(*buf) & 0xf];
      *dst++ = ' ';
   }

   *--dst = '\0';
   return len;
}


void log_hex(const char *buf, int len)
{
   char s[16 * 3];

   data_to_hex(s, buf, len > 16 ? 16 : len);
   log_msg(LOG_DEBUG, "data: %s", s);
}


void print_mac_table(if_info_t *ii)
{
   char buf[4096];

   snprint_mac_table(buf, sizeof(buf), &ii->mtbl);
   printf("===== %s =====\n%s\n", ii->ifname, buf);
}


void print_if_info(if_info_t *ii)
{
   char hwaddr[32], hwclient[32], hwrouter[32];

   ether_ntoa_r((struct ether_addr*) ii->hwaddr, hwaddr);
   pthread_mutex_lock(&ii->mutex);
   ether_ntoa_r((struct ether_addr*) ii->hwclient, hwclient);
   ether_ntoa_r((struct ether_addr*) ii->hwrouter, hwrouter);
   pthread_mutex_unlock(&ii->mutex);

   printf(
         "===== %s =====\n"
         "fd = %d\n"
         "wfd = %d\n"
         "filter = 0x%p\n"
         "out = %s\n"
         "gate = %s\n"
         "off = %d\n"
         "hwaddr = %s\n"
         "hwclient = %s\n"
         "hwrouter = %s\n\n"
         ,
         ii->ifname, ii->fd, ii->wfd, ii->filter, ii->out->ifname, ii->gate != NULL ? ii->gate->ifname : "NULL", ii->off, hwaddr, hwclient, hwrouter);
}


void cli_help(void)
{
   printf(
         "addr ......... List address tables.\n"
         "debug ........ Set debug level to DEBUG (7).\n"
         "exit ......... Exit program.\n"
         "info ......... Show interface info.\n"
         "nodebug ...... Set debug level to INFO (6).\n"
         "state ........ Show state table.\n"
         );
}


void cli(if_info_t *ii, int n)
{
   int running, i;
   char *s, *eptr;
   char buf[65536];

   printf("Welcome to %s!\n", PACKAGE_STRING);
   for (running = 1; running;)
   {
      printf("endoor# ");
      if (fgets(buf, sizeof(buf), stdin) == NULL)
         break;

      if ((s = strtok_r(buf, " \r\n", &eptr)) == NULL)
         continue;

      if (!strcmp(s, "exit"))
         running = 0;
      else if (!strcmp(s, "debug"))
         debug_level_ = 7;
      else if (!strcmp(s, "nodebug"))
         debug_level_ = 6;
      else if (!strcmp(s, "help"))
         cli_help();
      else if (!strcmp(s, "addr"))
      {
         for (i = 0; i < n; i++)
            print_mac_table(&ii[i]);
      }
      else if (!strcmp(s, "info"))
      {
         for (i = 0; i < n; i++)
            print_if_info(&ii[i]);
      }
      else if (!strcmp(s, "state"))
      {
         snprint_states(ii[2].st, buf, sizeof(buf));
         printf("%s\n", buf);
      }
   }
   printf("Good bye!\n");
}

